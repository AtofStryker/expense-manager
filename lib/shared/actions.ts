import { set } from '@siegrift/tsfunct'

import { Action } from '../redux/types'

export const setAppError = (error: string | null): Action => ({
  type: 'Set app error',
  reducer: (state) => set(state, ['error'], error),
})
