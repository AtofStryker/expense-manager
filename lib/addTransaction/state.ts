import { DEFAULT_CURRENCY } from '../shared/currencies'
import { ObjectOf } from '../types'

export enum TransactionTypes {
  fromUser = 'fromUser',
  imported = 'imported',
  repeated = 'repeated',
}

export type TransactionType = keyof typeof TransactionTypes

export interface Tag {
  id: string
  name: string
  uid: string
}

export enum RepeatingOptions {
  none = 'none',
  inactive = 'inactive',
  daily = 'daily',
  weekly = 'weekly',
  monthly = 'monthly',
  annually = 'annually',
}

export type RepeatingOption = keyof typeof RepeatingOptions

export interface BaseTransaction {
  transactionType: TransactionType
  // NOTE: order might be important
  tagIds: string[]
  currency: string
  isExpense: boolean
  note: string
  repeating: RepeatingOption
}

export interface Transaction extends BaseTransaction {
  id: string
  dateTime: Date
  amount: number
  uid: string
}

export interface AddTransaction extends BaseTransaction {
  useCurrentTime: boolean
  newTags: ObjectOf<Tag>
  dateTime?: Date
  tagInputValue: string
  amount: string
  shouldValidateAmount: boolean
}

export const createDefaultAddTransactionState = (): AddTransaction => ({
  transactionType: TransactionTypes.fromUser,
  amount: '',
  tagIds: [],
  newTags: {},
  currency: DEFAULT_CURRENCY.value,
  tagInputValue: '',
  isExpense: true,
  note: '',
  dateTime: undefined,
  useCurrentTime: true,
  shouldValidateAmount: false,
  repeating: RepeatingOptions.none,
})
