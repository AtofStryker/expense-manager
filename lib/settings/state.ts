import { DEFAULT_CURRENCY, Currency } from '../shared/currencies'
import { FirebaseField } from '../types'

export interface Settings {
  mainCurrency: Currency
  defaultCurrency: Currency
}
export interface ExchangeRates {
  rates: { [key in Currency]: number }
  base: Currency
  date: string
}

export interface Profile extends FirebaseField {
  settings: Settings
  exchangeRates: ExchangeRates
}

export const getInitialState = (user: firebase.User): Profile => ({
  uid: user.uid,
  id: user.uid,
  settings: {
    mainCurrency: DEFAULT_CURRENCY,
    defaultCurrency: DEFAULT_CURRENCY,
  },
  exchangeRates: {
    rates: {
      EUR: 1,
      CHF: 1.0784,
      HKD: 9.0882,
      SEK: 10.2958,
      USD: 1.1726,
      GBP: 0.90013,
      HUF: 345.72,
      CZK: 26.319,
      AUD: 1.6508,
      HRK: 7.4755,
      PLN: 4.4201,
    },
    base: 'EUR',
    date: '2020-08-03',
  },
})
