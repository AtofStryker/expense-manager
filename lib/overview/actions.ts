import { set } from '@siegrift/tsfunct'

import { Action } from '../redux/types'
import { OverviewPeriod } from '../state'

export const setOverviewPeriod = (
  period: OverviewPeriod,
): Action<OverviewPeriod> => ({
  type: 'Set overview period',
  payload: period,
  reducer: (state) => set(state, ['overviewPeriod'], period),
})
