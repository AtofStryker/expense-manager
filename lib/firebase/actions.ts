import { set } from '@siegrift/tsfunct'
import { batch } from 'react-redux'

import { addRepeatingTxs } from '../actions'
import { getFirebase } from '../firebase/firebase'
import { Action, Thunk } from '../redux/types'
import { SignInStatus } from '../state'
import { FirestoneQuery, getQueries } from './firestoneQueries'

export const firestoneChangeAction = (
  query: FirestoneQuery,
  payload: firebase.firestore.QuerySnapshot,
  isInitial = false,
): Action<firebase.firestore.QuerySnapshot> => ({
  type: `${isInitial ? 'Initial firestore' : 'Firestore'} query change: ${
    query.type
  }`,
  payload,
  reducer: (state) => {
    return query.reducer(state, payload)
  },
})

const changeSignInStatus = (status: SignInStatus): Action<SignInStatus> => ({
  type: 'Change sign in status',
  payload: status,
  reducer: (state) => set(state, ['signInStatus'], status),
})

export const authChangeAction = (status: SignInStatus): Thunk => async (
  dispatch,
  getState,
  { logger },
) => {
  logger.log(`Auth changed: ${status}`)
  if (status === 'loggedIn') {
    await dispatch(initializeFirestore())
    // TODO: remove or fix repeating transactions
    // dispatch(addRepeatingTxs())
  }
  dispatch(changeSignInStatus(status))
}

export const initializeFirestore = (): Thunk => async (dispatch) => {
  const queries = getQueries()
  const initialQueries = queries.map((query) => {
    return query.createFirestoneQuery().get({
      source: 'cache',
    })
  })

  const initialQueriesData = await Promise.all(initialQueries)
  batch(() => {
    initialQueriesData.forEach((data, i) =>
      dispatch(firestoneChangeAction(queries[i], data, true)),
    )
  })

  let actions: Array<Parameters<typeof firestoneChangeAction>> = []
  getFirebase()
    .firestore()
    .onSnapshotsInSync(() => {
      // https://react-redux.js.org/api/batch
      // treat the redux updates as one atomic operation and forbid rendering between the updates
      // (which can render transaction with tag that hasn't been loaded yet)
      batch(() => {
        actions.forEach((a) => dispatch(firestoneChangeAction(a[0], a[1])))
        actions = []
      })
    })

  queries.forEach((q) => {
    q.createFirestoneQuery().onSnapshot((change) => {
      actions.push([q, change])
    })
  })
}
