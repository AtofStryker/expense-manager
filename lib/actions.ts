import { set } from '@siegrift/tsfunct'
import { firestore } from 'firebase/app'
import { chunk } from 'lodash'

import { Tag, Transaction } from './addTransaction/state'
import { Action, Thunk } from './redux/types'
import { ScreenTitle } from './state'
import { ObjectOf } from './types'

export const setCurrentScreen = (screen: ScreenTitle): Action<ScreenTitle> => ({
  type: 'Set current screen',
  payload: screen,
  reducer: (state) => set(state, ['currentScreen'], screen),
})

export const uploadTransactions = (txs: Transaction[]): Thunk => (
  dispatch,
  getState,
  { logger },
) => {
  logger.log('Upload multiple transactions to firestone')

  const coll = firestore().collection('transactions')
  const uploads = chunk(txs, 500).map((ch) => {
    const batch = firestore().batch()
    ch.forEach((tx) => {
      const ref = coll.doc(tx.id)
      batch.set(ref, tx)
    })

    return batch.commit()
  })

  return Promise.all(uploads)
}

export const uploadTransaction = (tx: Transaction): Thunk => (
  dispatch,
  getState,
  { logger },
) => {
  logger.log('Upload transaction to firestone')

  return firestore()
    .collection('transactions')
    .doc(tx.id)
    .set(tx)
    .catch((error) => {
      // TODO: handle errors
      console.error('Error writing new message to Firebase Database', error)
    })
}

export const uploadTags = (tags: ObjectOf<Tag>): Thunk => (
  dispatch,
  getState,
  { logger },
) => {
  logger.log('Upload tags to firestone')

  const coll = firestore().collection('tags')
  const uploads = chunk(Object.values(tags), 500).map((ch) => {
    const batch = firestore().batch()
    ch.forEach((tag) => {
      const ref = coll.doc(tag.id)
      batch.set(ref, tag)
    })

    return batch.commit()
  })

  return Promise.all(uploads)
}
