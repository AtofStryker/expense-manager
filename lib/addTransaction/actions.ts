import { omit } from '@siegrift/tsfunct'
import uuid from 'uuid/v4'

import { uploadToFirebase } from '../actions'
import { getCurrentUserId } from '../firebase/util'
import { Thunk } from '../redux/types'

import { AddTransaction } from './state'

export const addTransaction = (addTx: AddTransaction): Thunk => async (
  dispatch,
  getState,
  { logger },
) => {
  logger.log('Add transaction')

  const id = uuid()
  const tx = {
    id,
    ...omit(addTx, ['newTags', 'tagInputValue', 'useCurrentTime']),
    amount: Number.parseFloat(addTx.amount),
    dateTime: addTx.useCurrentTime ? new Date() : addTx.dateTime!,
    uid: getCurrentUserId(),
  }

  await dispatch(
    uploadToFirebase([tx], Object.values(getState().addTransaction.newTags)),
  )
}
