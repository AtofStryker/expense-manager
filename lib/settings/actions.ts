import { keyBy } from 'lodash'
import uuid from 'uuid/v4'

import { uploadTags, uploadTransactions } from '../actions'
import { Tag, Transaction } from '../addTransaction/state'
import firebase from '../firebase/firebase'
import { Action, Thunk } from '../redux/types'
import { currencies } from '../shared/currencies'
import { downloadFile, isValidDate } from '../shared/utils'
import { State } from '../state'
import { ObjectOf } from '../types'

import { exportedCsvSel } from './selectors'

export const importFromCSV = (
  e: React.ChangeEvent<HTMLInputElement>,
): Thunk => (dispatch, getState, { logger }) => {
  logger.log('Import from csv')
  const chosenFile = e.target.files!.item(0)!
  const reader = new FileReader()

  return new Promise((res) => {
    reader.onload = async () => {
      const { errorReason, tags, txs } = processImportedCSV(
        getState(),
        reader.result as string,
      )

      if (errorReason) {
        // TODO: add to state and display notification
        console.log(errorReason)
      } else {
        // TODO: display success notification
        await dispatch(uploadTags(keyBy([...tags.values()], 'id')))
        await dispatch(uploadTransactions(txs))
      }

      res()
    }

    reader.readAsText(chosenFile)
  })
}

/**
 * Imported csv file must have the following column structure:
 *  1. date - in ISO 8601 format
 *  2. amount - 2 decimal places, negative values means expenses, positive are incomes
 *  3. tags - separated by '|' (pipe)
 *  4. note - optional
 *  5. currency - currency value of transaction (see list of available currency values)
 *  other columns are ignored
 */
export const processImportedCSV = (state: State, importedCsv: string) => {
  const lines = importedCsv.trim().split('\n')
  const txs: Transaction[] = []
  const tags = new Map<string, Tag>()
  const stateTagsByName = Object.values(state.tags).reduce(
    (acc, tag) => ({ ...acc, [tag.name]: tag }),
    {} as ObjectOf<Tag>,
  )
  let errorReason: string | undefined

  try {
    for (const line of lines) {
      const t = line.split(',')

      // check for validity
      const dt = new Date(t[0])
      const rawTags = t[2].trim()
      if (!isValidDate(dt)) {
        errorReason = `${t[0]} is not a valid date`
      } else if (t[1].match(/^-?\d+(\.\d{1,2})?$/) === null) {
        errorReason = `${t[1]} is not in a valid amount format`
      } else if (rawTags.length === 0) {
        errorReason = `There must be at least one tag in a transaction`
      } else if (!currencies.find((c) => c.value === t[4])) {
        errorReason = `Invalid currency of a transaction`
      }

      // there mustn't be any error when importing
      if (errorReason) {
        break
      }

      const amount = Number.parseFloat(t[1])
      const splitTags = rawTags.split('|')
      splitTags.forEach((tag) => {
        if (!stateTagsByName[tag] && !tags.has(tag)) {
          tags.set(tag, { id: uuid(), name: tag })
        }
      })
      txs.push({
        id: uuid(),
        amount: Math.abs(amount),
        transactionType: 'imported',
        currency: t[4],
        dateTime: dt,
        isExpense: amount < 0,
        note: t[3],
        tagIds: splitTags.map((tag) => {
          if (tags.get(tag)) {
            return tags.get(tag)!.id
          } else {
            return stateTagsByName[tag].id
          }
        }),
      })
    }
  } catch (e) {
    errorReason =
      'Unknown error. Check whether the data are in correct format.'
  }

  return { errorReason, txs, tags }
}

export const exportToCSV = (): Action => ({
  type: 'Export to csv',
  reducer: (state) => {
    downloadFile('expense-manager-data.csv', exportedCsvSel(state))
    return state
  },
})

export const clearAllData = (): Thunk => async (
  dispatch,
  getState,
  { logger },
) => {
  logger.log('Clear all data')
  const removeColl = async (name: string) => {
    let stopRemove = false
    while (!stopRemove) {
      const batch = firebase.firestore().batch()

      const q = await firebase
        .firestore()
        .collection(name)
        .limit(500)
        .get()

      if (q.size > 0) {
        q.docs.forEach((d) => batch.delete(d.ref))
        batch.commit()
      } else {
        stopRemove = true
      }
    }
  }
  return Promise.all([removeColl('transactions'), removeColl('tags')])
}
