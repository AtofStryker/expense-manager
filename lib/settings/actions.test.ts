import { getInitialState, State } from '../state'

import { processImportedCSV } from './actions'

jest.mock('uuid/v4', () => {
  let uuid = 0
  return jest.fn(() => `uuid${uuid++}`)
})

describe('settings actions', () => {
  describe('import/export', () => {
    describe('processing imported txs', () => {
      let state: State

      beforeEach(() => {
        state = getInitialState()
      })

      test('checks for errors', () => {
        const checkError = (input: string) =>
          processImportedCSV(state, input).errorReason

        expect(checkError('')).toBe(
          'Unknown error. Check whether the data are in correct format.',
        )

        expect(checkError('invalid_date,-1,Travel|Tickets')).toBe(
          'invalid_date is not a valid date',
        )
        expect(checkError('21.11.1999,-1,Travel|Tickets')).toBe(
          '21.11.1999 is not a valid date',
        )
        expect(checkError('2017.08.19T00:00:00Z,-1,Travel|Tickets')).toBe(
          '2017.08.19T00:00:00Z is not a valid date',
        )

        expect(checkError('2017-08-19T00:00:00Z,+2,Travel|Tickets')).toBe(
          '+2 is not in a valid amount format',
        )
        expect(checkError('2017-08-19T00:00:00Z,-1.345,Travel|Tickets')).toBe(
          '-1.345 is not in a valid amount format',
        )

        expect(checkError('2017-08-19T00:00:00Z,2,')).toBe(
          'There must be at least one tag in a transaction',
        )
      })

      test('creates new tags and txs', () => {
        state.tags = { id1: { id: 'id1', name: 'Travel' } }
        const csv1 =
          '2017-08-19T00:00:00Z,-1,Travel|Tickets,\n2017-08-16T00:00:00Z,-0.7,Shopping,'
        const csv2 = '2016-02-16T00:00:00Z,-1.56,A|B|C,'

        // Travel tag should be reused from state
        expect(processImportedCSV(state, csv1)).toMatchSnapshot()
        expect(processImportedCSV(state, csv2)).toMatchSnapshot()
      })
    })
  })
})
