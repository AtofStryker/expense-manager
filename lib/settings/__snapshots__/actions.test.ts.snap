// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`settings actions import/export processing imported txs creates new tags and txs 1`] = `
Object {
  "errorReason": undefined,
  "tags": Map {
    "Tickets" => Object {
      "id": "uuid0",
      "name": "Tickets",
    },
    "Shopping" => Object {
      "id": "uuid2",
      "name": "Shopping",
    },
  },
  "txs": Array [
    Object {
      "amount": 1,
      "currency": "EUR",
      "dateTime": 2017-08-19T00:00:00.000Z,
      "id": "uuid1",
      "isExpense": true,
      "note": "",
      "tagIds": Array [
        "id1",
        "uuid0",
      ],
      "transactionType": "imported",
    },
    Object {
      "amount": 0.7,
      "currency": "EUR",
      "dateTime": 2017-08-16T00:00:00.000Z,
      "id": "uuid3",
      "isExpense": true,
      "note": "",
      "tagIds": Array [
        "uuid2",
      ],
      "transactionType": "imported",
    },
  ],
}
`;

exports[`settings actions import/export processing imported txs creates new tags and txs 2`] = `
Object {
  "errorReason": undefined,
  "tags": Map {
    "A" => Object {
      "id": "uuid4",
      "name": "A",
    },
    "B" => Object {
      "id": "uuid5",
      "name": "B",
    },
    "C" => Object {
      "id": "uuid6",
      "name": "C",
    },
  },
  "txs": Array [
    Object {
      "amount": 1.56,
      "currency": "EUR",
      "dateTime": 2016-02-16T00:00:00.000Z,
      "id": "uuid7",
      "isExpense": true,
      "note": "",
      "tagIds": Array [
        "uuid4",
        "uuid5",
        "uuid6",
      ],
      "transactionType": "imported",
    },
  ],
}
`;
